Fix path expansion for traces when log_root is set

hcs42 opened the pull request:
Previously the following configuration didn't work properly:

    {log_root, "logs"},
    {handlers, [{lager_file_backend, [{file, "access.log"}, {level, none}, ...]}]},
    {traces, [{{lager_file_backend, "access.log"}, [{tag, access}], info}]}

When lager:trace_file was trying to find a file among the existing handlers, it was searching for the expanded file name ("logs/access.log"), but the handlers store the file names without expansion ("access.log"), so it didn't find it.

The result was that lager:trace_file started another "access.log" backend, which of course didn't use e.g. the log rotation settings of the first handler. When the log_root was relative, then each log message appeared twice in the log file (both backends thought that they should log it, because when calculating the Destination, "logs/access.log" was used in both cases).

Now the code performs path expansion on the file name inside the handler too. Also, now it calculates the absolute paths of both files, so that if one file is specified with an absolute path and the other with a relative path, lager will still know that they are the same.

mrallen1 commented:
Thanks, this is super dope. I have one small request, though. Would you please add a few test cases for your code so we can be sure we don't regress on this in the future?

hcs42 commented:
Sure!

hcs42 commented:
Update: witing the automatic test is in progress (it's a bit more tricky than I expected :) )

mrallen1 commented:
It can be. If you're stuck, I'd be happy to help point you in the right direction. Thanks so much for your contribution.

hcs42 commented:
Sorry, I've been busy with other things... I will look at this on the weekend and either finish the test (and rebase it :) ) or ask for some pointers.

mrallen1 commented:
Thanks!

hcs42 commented:
I have added the test and force pushed the commit.

The test contains a timer:sleep(1000) call, because the error with the old code is that *eventually* the log file will contain the log messag twice. My idea for testing this *eventually* is to sleep one second and check if duplication has happened. If there is a better way, I'm open to suggestions.

mrallen1 commented:
This is great. Curious why you didn't enclose the timer in the wait_for function you quoted - that seems like a better way to handle that "wait for" condition than a blanket sleep. If you want to adjust the time between the first call and the second, just make the sleep time a parameter you pass into it.

hcs42 commented:
Thanks for the tips!

> Curious why you didn't enclose the timer in the wait_for function you quoted - that seems like a better way to handle that "wait for" condition than a blanket sleep.

Simply enclosing the timer in wait_for (or wait_until, as it's more often seems to be called in the codebase) would make it easy for the test to pass even when it is buggy. After the first line is written into the file, wait_for could quickly conclude that this is fine, test passed. But it may happen that the second line is written into the file a little later. (This is what I usually saw in the production log files.) My solution was make sure that we give time to the system for the second line to arrive, hence timer:sleep(1000). I could add a wait_for call after the timer:sleep(1000), so that if the first line arrives later than a second, we still pass.

> If you want to adjust the time between the first call and the second, just make the sleep time a parameter you pass into it.

If we want to replace timer:sleep with wait_for, we need to make sure that the sleep happens first (currently wait_until is happy to continue quickly if the condition holds). So we would need to move the timer:sleep inside wait_for to the beginning.

What do you think, does it make sense to add a wait_for after the timer:sleep(1000), or to create a new wait_for with the sleep parameter and moving the timer:sleep to the beginning?

mrallen1 commented:
I would wrap the rest of the test (after waiting for the file to exist) in another wait_for function. I would simplify the output of lager's messages using a custom formatting function. Something like

```erlang
format(Msg, _Config) -> lager_msg:message(Msg)
```

should get you the "bare" message contents without timestamps or severity, etc. You can use this format/2 function as a "custom formatter" when you set up your test file in the test setup.  (See [Custom Formatting](https://github.com/erlang-lager/lager#custom-formatting) in the README.)

Then I would repeatedly read the log file contents, split it into lines with something like

```erlang
binary:split(FileContents, <<"\n">>, [global, trim_all])
```

and inspect the list that results to see if your log message has been duplicated. 

If not, then wait some more. If the log message does not show up as duplicated after the wait_for times out, then pass the test, otherwise, fail the test if a duplicate message shows up.

Since this test might run for 2 or 3 seconds, it would be a good idea to wrap the whole thing in an extended timeout tuple for EUnit.

```
{timeout, 10, fun do_test/0}
```

or whatever works best.  Running unit tests that go quick on your own workstation take 2 or 3 times longer on TravisCI and can lead to spurious test failures.

mrallen1 commented:
During our lager PR triage today, @jlouis suggested a function like the one in this gist might be a good way to build a test that waits for a condition with a sleep that backs off exponentially

https://gist.github.com/anonymous/a44fefd9dbf6d610b2be9aa7ec7079d0

Would you have time to incorporate something like that in your test?

hcs42 commented:
Thank you. I will do that on the weekend.

hcs42 commented:
Thank you for all your suggestions (and patience...)

I incorporated all of them. Feel free to modify it before merging, especially the timeout values.

* I use another `wait_until` function to check for the duplicate log entry.
* I use lager's formatter to make the message minimal.
* I use binary:split on the file contents. I actually only examine the number of lines when everything looks ok.
* I wrapped the unit test into a timeout tuple.
* I used jlouis's gist to back off according to the Fibonacci sequence.

hcs42 commented:
Note: I used `trim` instead of `trim_all` in `binary:split`. The latter works only from Erlang 18, which made travis unhappy.

mrallen1 commented:
Looks really good. I just got back from traveling, so I'm going to wait until tomorrow to merge but, this is fantastic. 

Thank you!

mrallen1 commented:
Just ran everything locally to confirm. This looks good. Thanks!!
